astropy/io/ascii/tests/test_c_reader.py

```diff
@@ -1110,12 +1125,146 @@ def test_fortran_invalid_exp(parallel):
+    # tested entries and expected returns, first for auto-detect,
+    # then for different specified exponents
+    fields = [ '1.0001+1', '.42d1', '2.3+10', '0.5', '3+1001', '3000.',
+               '2', '4.56e-2.3', '8000', '4.2-022', '.00000145e314' ]
+    vals_e = [ '1.0001+1', '.42d1', '2.3+10',   0.5, '3+1001',  3.e3,
+               2, '4.56e-2.3',    8000,  '4.2-022', 1.45e308 ]
+    vals_d = [ '1.0001+1',     4.2, '2.3+10',   0.5, '3+1001',  3.e3,
+               2, '4.56e-2.3',    8000,  '4.2-022', '.00000145e314' ]
+    vals_a = [ '1.0001+1',     4.2, '2.3+10',   0.5, '3+1001',  3.e3,
+               2, '4.56e-2.3',    8000,   4.2e-22,  1.45e308 ]
+    vals_v = [ '1.0001+1', 4.2, '2.3+10',   0.5, '3+1001',  3.e3,
+               2, '4.56e-2.3',    8000,  '4.2-022', 1.45e308 ]
+
+    # iterate over supported format types and separators
+    for f, s in formats.items():
+        t1 = ascii.read(StringIO(s.join(header)+'\n'+s.join(fields)),
+                        format=f, guess=guess,
+                        fast_reader={'parallel': parallel, 'exponent_style': 'A'})
+        # assert t1['I1'].dtype.kind == 'i'
+        assert_table_equal(t1, Table([[col] for col in vals_a], names=header))
+
+    # try another separator as well with auto-detection
+    #formats['bar'] = '|'
+
+    # non-basic separators require guessing enabled to be detected
+    if not guess:
+        formats = { 'basic': ' ' }
+
+    for s in formats.values():
+        t2 = ascii.read(StringIO(s.join(header)+'\n'+s.join(fields)), guess=guess,
+                fast_reader={'parallel': parallel, 'exponent_style': 'a'})
+
+        assert_table_equal(t2, Table([[col] for col in vals_a], names=header))
+
+    # iterate for (default) expchar 'E'
+    for s in formats.values():
+        t3 = ascii.read(StringIO(s.join(header)+'\n'+s.join(fields)), guess=guess,
+                fast_reader={'parallel': parallel, 'use_fast_converter': True})
+
+        assert_table_equal(t3, Table([[col] for col in vals_e], names=header))
+
+    # iterate for expchar 'D'
+    for s in formats.values():
+        t4 = ascii.read(StringIO(s.join(header)+'\n'+s.join(fields)), guess=guess,
+                fast_reader={'parallel': parallel, 'exponent_style': 'D'})
+
+        assert_table_equal(t4, Table([[col] for col in vals_d], names=header))
+
+    # iterate for regular converter (strtod)
+    for s in formats.values():
+        t5 = ascii.read(StringIO(s.join(header)+'\n'+s.join(fields)), guess=guess,
+                fast_reader={'parallel': parallel, 'use_fast_converter': False})
+
+        read_values = [col[0] for col in t5.itercols()]
+        if os.name == 'nt':
+            # apparently C strtod() on (some?) MSVC recognises 'd' exponents!
+            assert read_values == vals_v or read_values == vals_e
+        else:
+            assert read_values == vals_e
+
+
+def test_fortran_reader_notbasic():
+    """
+    Check if readers without a fast option raise a value error when a
```

Comment:
The first few tests seem unrelated to this description.  Can you add inline comments describing what each test is doing?